src/Microsoft.TemplateEngine.Utils/SemanticVersion.cs

```diff
@@ -0,0 +1,438 @@
+            {
+            }
+
+            //If the version part was empty or multiple digits that started with zero,
+            //  this isn't a valid version string
+            if (majorEnd == 0 || (source[0] == '0' && majorEnd > 1))
+            {
+                version = null;
+                return false;
+            }
+
+            if (majorEnd == source.Length || source[majorEnd] != '.')
+            {
+                tail = majorEnd;
+                goto HandlePostVersion;
+            }
+
+            for (minorEnd = majorEnd + 1; minorEnd < source.Length && source[minorEnd] >= '0' && source[minorEnd] <= '9'; ++minorEnd)
+            {
+            }
+
+            if (minorEnd == majorEnd + 1 || (source[majorEnd + 1] == '0' && minorEnd - majorEnd - 1 > 1))
+            {
+                version = null;
+                return false;
+            }
+
+            if (minorEnd == source.Length || source[minorEnd] != '.')
+            {
+                tail = minorEnd;
+                goto HandlePostVersion;
+            }
+
+            for (patchEnd = minorEnd + 1; patchEnd < source.Length && source[patchEnd] >= '0' && source[patchEnd] <= '9'; ++patchEnd)
+            {
+            }
+
+            if (patchEnd == minorEnd + 1 || (source[minorEnd + 1] == '0' && patchEnd - minorEnd - 1 > 1))
+            {
+                version = null;
+                return false;
+            }
+
+            tail = patchEnd;
+
+HandlePostVersion:
+            int major = int.Parse(source.Substring(0, majorEnd), NumberStyles.None, CultureInfo.InvariantCulture);
+            int minor = minorEnd > majorEnd ? int.Parse(source.Substring(majorEnd + 1, minorEnd - majorEnd - 1), NumberStyles.None, CultureInfo.InvariantCulture) : 0;
+            int patch = patchEnd > minorEnd ? int.Parse(source.Substring(minorEnd + 1, patchEnd - minorEnd - 1), NumberStyles.None, CultureInfo.InvariantCulture) : 0;
+            string prerelease = null;
+            string metadata = null;
+
+            if (tail < source.Length && source[tail] == '-')
+            {
+                int end = source.IndexOf('+', tail);
+
+                if (end < 0)
+                {
+                    end = source.Length;
+                }
+
+                if (tail == end - 1)
+                {
+                    version = null;
+                    return false;
+                }
+
+                for (int i = tail + 1; i < end; ++i)
+                {
+                    if (!(
+                        source[i] == '-' || source[i] == '.' ||
+                        (source[i] >= 'A' && source[i] <= 'Z') ||
+                        (source[i] >= 'a' && source[i] <= 'z') ||
+                        (source[i] >= '0' && source[i] <= '9')
+                        ))
+                    {
+                        version = null;
+                        return false;
+                    }
+                }
```

Comment:
Per the spec, there should really be a check here that disallows period separated segments that consist only of digits that begin with a 0 that are more than one character in length